File: projects/blognami/lib/blognami/models/tagable.rb

require "slick/database/row"

Slick::Database::Row.register "tagable", abstract: true do

  has_many :tagable_tags, from_key: :id, to_key: :tagable_id
  has_many :tags, through: [:tagable_tags, :tag]

  scope :tagged_with do |*tag_names|
    out = self
    tag_names.each do |tag_name|
        out = out.tags.name_eq(tag_name).back(2)
    end
    out
  end

  after_insert_or_update do
    next if @tags.nil?
    tags = @tags.to_s.split(/\n/).map{|tag| tag.strip.gsub(/\s+/, ' ')}.select{|tag| tag != ''}
    tag_ids = []
    tags.each do |name|
        tag = database.tags.name_eq(name).first
        if tag
            tag_ids << tag.id
        else
            tag_ids << database.tags.insert(name: name).id
        end
    end

    tagable_tags.delete

    tag_ids.each do |tag_id|
        database.tagable_tags.insert(tagable_id: id, tag_id: tag_id)
    end
  end

  def tags=(value)
    @tags = value
  end

end

File: packages/@blognami/main/lib/models/tagable.js

export default {
    meta(){
        this.hasMany('tagableTags', { fromKey: 'id', toKey: 'tagableId' });
        this.hasMany('tags', { through: ['tagableTags', 'tag'] });

        this.afterInsertOrUpdate(async function(){
            if(typeof this._tags != 'string') return;
            const tags = this._tags.split(/\n/).map(tag => tag.trim().replace(/\s+/g, ' ')).filter(tag => tag != '');
            const tagIds = [];
            while(tags.length > 0){
                const name = tags.shift();
                const tag = await this.database.tags.where({ name }).first();
                if(tag){
                    tagIds.push(tag.id)
                } else {
                    tagIds.push((await this.database.tags.insert({ name })).id);
                }
            }

            await this.tagableTags.delete();
            
            while(tagIds.length){
                await this.database.tagableTags.insert({
                    tagableId: this.id,
                    tagId: tagIds.shift()
                });
            }
        });

        this.scope('taggedWith', function(tagNames){
            [ tagNames ].flat().forEach(tagName => this.tags.where({ name: tagName }));
        });
    },

    set tags(tags){
        this._tags = tags;
    }
};
